UIforETW/Utility.cpp

```diff
@@ -126,71 +459,126 @@ std::wstring ConvertToCRLF(const std::wstring& input)
 	std::vector<wchar_t> buffer(length + 1);
-	GetWindowText(hEdit, &buffer[0], static_cast<int>(buffer.size()));
+	
+	//GetWindowText https://msdn.microsoft.com/en-us/library/windows/desktop/ms633520.aspx
+	//If [GetWindowTextW] succeeds, the return value is the length,
+	//in characters, of the copied string, not including the
+	//terminating null character.
+
+	//If the window has no title bar or text,
+	//[or] if the title bar is empty,
+	//or if the window or control handle is invalid,
+	//the return value is zero. 
+	
+	const int charsWritten = ::GetWindowTextW(hEdit, &buffer[0], static_cast<int>(buffer.size()));
+	if (charsWritten == 0)
+	{
+		if (length > 0)
+			debugLastError();
+		return L"";
+	}
+	UIETWASSERT(charsWritten == length);
+	UIETWASSERT((charsWritten) < (int)buffer.size());
+	UIETWASSERT(buffer[charsWritten] == 0);
 	// Double-verify that the buffer is null-terminated.
 	buffer[buffer.size() - 1] = 0;
 	return &buffer[0];
 }
 
+int RequiredNumberOfWideChars(const std::string& text)
+{
+	static_assert( sizeof(std::string::value_type) == 1 == sizeof(text[0]),
+		"bad assumptions!");
+
+	const int multiCharCount = ::MultiByteToWideChar(CP_ACP, 0, text.c_str(),
+		static_cast<int>(text.size() + 1), NULL, 0);
+
+	if (multiCharCount == 0)
+	{
+		//No reasonable way for MultiByteToWideChar to fail.
+		debugLastError( );
+		std::terminate( );
+	}
+	return multiCharCount;
+}
+
 std::wstring AnsiToUnicode(const std::string& text)
 {
+	//if the string is empty, then we can return early, and avoid
+	//confusing return values (from MultiByteToWideChar)
+	if (text.empty())
+		return L"";
+
 	// Determine number of wide characters to be allocated for the
 	// Unicode string.
-	size_t cCharacters = text.size() + 1;
+	const int multiCharCount = RequiredNumberOfWideChars(text);
 
-	std::vector<wchar_t> buffer(cCharacters);
+	std::vector<wchar_t> buffer(multiCharCount);
 
 	// Convert to Unicode.
-	std::wstring result;
-	if (MultiByteToWideChar(CP_ACP, 0, text.c_str(), static_cast<int>(cCharacters), &buffer[0], static_cast<int>(cCharacters)))
+	const int multiToWideResult = ::MultiByteToWideChar(CP_ACP, 0, text.c_str(), static_cast<int>(text.size() + 1), &buffer[0], multiCharCount);
+	if (multiToWideResult == 0)
 	{
-		// Double-verify that the buffer is null-terminated.
-		buffer[buffer.size() - 1] = 0;
-		result = &buffer[0];
-		return result;
+		//No reasonable way for MultiByteToWideChar to fail.
+		debugLastError( );
+		std::terminate( );
 	}
 
+	UIETWASSERT(multiToWideResult > 0);
+	UIETWASSERT(buffer[multiToWideResult - 1] == 0);
+
+	// Double-verify that the buffer is null-terminated.
+	buffer[buffer.size() - 1] = 0;
+	std::wstring result = &buffer[0];
 	return result;
 }
 
 // Return a string from a format string and some printf-style arguments.
-std::wstring stringPrintf(_Printf_format_string_ const wchar_t* pFormat, ...)
+std::wstring stringPrintf(_Printf_format_string_ PCWSTR const pFormat, ...)

```

Comment:
We actually [discussed this in a previous PR (the "ETW providers cleanup")](https://github.com/google/UIforETW/pull/33#discussion_r30299346), and I tried to set up a test for this the other day, but `/analyze` was insanely flaky. It refused to even pick up obvious non-null terminated strings. If you're curious, or just need a reason to be angry at /analyze, you can check it out here: https://github.com/ariccio/test-printf-format-string-annotation/